Base Docker image for OE

CyanDevs opened the pull request:
This creates a base Docker image that will provide the minimum packages required to run OE.

To build the Docker image, use build.sh
```
Usage: ../build.sh -v <string> -i <string> [OPTIONS..]
Create a base Docker image for Open Enclave
  -v     Open Enclave release version (Example: 0.17.2)
  -i     Intel SGX version (Example: 2.15.100)

Options:

  -u     Ubuntu release version [Default: 18.04]
  -t     Tag for the Docker image [Default: latest]
```

An example command to build would be
```
../build.sh -v 0.17.2 -i 2.15.100 -u 20.04 -t 0.17.2
```
where
`-v` specifies version 0.17.2 of OE
`-i` specifies version 2.15.100 of Intel SGX
`-u` specifies Ubuntu release version 20.04 (Focal) (This is optional -- defaults to 18.04)
`-t` specifies the tag "0.17.2" for the Docker image

This results in the following Docker images (built for both Ubuntu 20.04 and 18.04):
```
REPOSITORY           TAG       IMAGE ID       CREATED              SIZE
openenclave-bionic   0.17.2    7e5ec5824bb3   About a minute ago   462MB
openenclave-focal    0.17.2    38c99b884f32   2 minutes ago        327MB
```

Note that in 20.04, default build tools are not available and can be installed by running
```
apt update
apt install -y build-essential
```

Signed-off-by: Chris Yan <chrisyan@microsoft.com>

anakrish commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+# Install Intel SGX and Open Enclave
+RUN apt-get update && \
+    apt-get install -y \
+      libssl-dev \
```
Can you double-check if the dev package is needed to run OE? Typically, dev packages are only needed to build applications.

CyanDevs replied:
Yes you are correct, it is needed to build, but not needed to run. I'll remove this. Thanks!

anakrish commented:
Thanks for doing this! This will be very useful.

anakrish commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+# Register required apt repositories
+RUN echo "deb [arch=amd64 signed-by=/usr/share/keyrings/intel-sgx-keyring.gpg] https://download.01.org/intel-sgx/sgx_repo/ubuntu ${UBUNTU_CODENAME} main" | tee /etc/apt/sources.list.d/intel-sgx.list
+
+# Install Intel SGX and Open Enclave
```
```suggestion
# Install Intel SGX and Azure DCAP client
```

CyanDevs commented:
Latest changes:
* Removed package libssl-dev
* Removed package open-enclave and all associated code with the versioning
* Minor comment changes
* Changed the installed packages list to be more explicit in what packages are installed.

anakrish commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+#
+# Please use the associated build.sh to build this Dockerfile
+ARG UBUNTU_CODENAME
+FROM ubuntu:${UBUNTU_CODENAME}
```
Does ubuntu have a minimal image? If so, that is preferable for running enclaves.

anakrish commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+RUN echo "deb [arch=amd64 signed-by=/usr/share/keyrings/msprod-keyring.gpg] https://packages.microsoft.com/ubuntu/${UBUNTU_VERSION}/prod ${UBUNTU_CODENAME} main" | tee /etc/apt/sources.list.d/msprod.list
+
+# Install Intel SGX and Azure DCAP Client
+RUN apt-get update && \
```
The container does not have build tools like gcc right?

anakrish replied:
Sounds good. Thanks for confirming.

radhikaj reviewed (approved):
Thanks for doing this.
We would need another container for devs to be able to build their enclave applications. But we can do that next.

rs-- commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/build.sh:
```diff
@@ -0,0 +1,119 @@
+echo "Checking for Intel SGX version ${SGX_VERSION} for Ubuntu ${UBUNTU_CODENAME}..."
+if [[ -d "${BUILD_DIR}/apt_preference_files" ]]; then
+    rm -rf "${BUILD_DIR}/apt_preference_files"
+    mkdir apt_preference_files
```
for consistency, should this be:
`mkdir "${BUILD_DIR}/apt_preferences_files"`

rs-- commented on .jenkins/infrastructure/docker/dockerfiles/linux/base/build.sh:
```diff
@@ -0,0 +1,119 @@
+if [[ -f "${BUILD_DIR}/intel-sgx-deb.key" ]]; then
+    rm "${BUILD_DIR}/intel-sgx-deb.key"
+fi
+wget \
```
Can the download+adding they key be put in a function since it's used in a couple of places?
